### ISSUE TITLE
Optimistic response not being applied after server response with multiple mutations

### ISSUE THREAD

**Issue by KennyHammerlund** [apollographql/apollo-client#7341]
**Intended outcome:**
Use optimistic response without it "flashing back' to old values

**Actual outcome:**
When firing off multiple mutations the optimistic updates are not alway being re-applied when the server returns a value for one mutation  

**How to reproduce the issue:**
This points to a local API but you can get an idea of settings. The api introduces a delay of 2500-5000ms to simulate poor network conditions. If you cant reproduce in a similar manner and want to use this sandbox send me a message and we can do a screenshare or something. 
https://codesandbox.io/s/apa-apollo-client-1lcko?file=/src/RackTurn.js

Originally I thought it had to do with 2 different mutations accessing the same entity but now it looks like I can reproduce the behavior by firing the same mutation multiple times.  

There is a button that has a mutation that adds a defensive shot. I can reproduce most of the time by clicking the button 3 times quickly. 

I have a console log in the update function that says optimistic when its an optimistic: `update DS..`
There is a useEffect on the screen to monitor the button changing values as it happens quick so leaving a record helps. 

This is an image of what I think is expected behavior (sometimes things work sometimes they dont).  
* With 3 requests outstanding when 1/3 comes back the update functions for 2/3 & 3/3 are applied with the optimistic values. 
* Then when 2/3 comes back from the server 3/3 update is ran with optimistic values. 
* finally the server responds with 3/3 and no optimistic values are applied. 
![image](https://user-images.githubusercontent.com/23132999/99412246-2255ab80-28ba-11eb-8f0f-4dffd0ee8ed6.png)

This is the same sequence with the bug. 
* when 2/3 comes back from the server with 199 the optimistic value/update to set it to 200 is never applied.
* since the optimistic 200 is not applied the screen "flashes back" from 200 to 199 
* finally when the server returns 200 the screen is updated correctly

![image](https://user-images.githubusercontent.com/23132999/99412827-c2abd000-28ba-11eb-957d-5ded8a7a15de.png)

**Versions**
  System:
    OS: macOS 10.15.7
  Binaries:
    Node: 12.18.3 - ~/.nvm/versions/node/v12.18.3/bin/node
    Yarn: 1.22.10 - ~/apa/scorekeeping/node_modules/.bin/yarn
    npm: 6.14.6 - ~/.nvm/versions/node/v12.18.3/bin/npm
  Browsers:
    Chrome: 86.0.4240.198
    Safari: 13.1.3
  npmPackages:
    @apollo/client: 3.2.5 => 3.2.5 
    @apollo/link-batch-http: 2.0.0-beta.3 => 2.0.0-beta.3 
    // The codesandbox reproduces the issue without the packages below
    @apollo/link-ws: 2.0.0-beta.3 => 2.0.0-beta.3 
    @apollo/react-testing: 3.1.4 => 3.1.4 
    apollo: 2.28.0 => 2.28.0 
    apollo-cache-persist: 0.1.1 => 0.1.1 
    apollo-link-offline: 1.0.0 => 1.0.0 
    apollo-link-retry: 2.2.16 => 2.2.16 
    apollo-link-serialize: ^3.1.1 => 3.1.1

**Comment by KennyHammerlund** [comment]
Upon further investigation I don't believe this is a bug. I think there is no better way to do Optimistic UI without implementing some custom logic for each app. maybe a type policy for mutation response payload would do it, something to explore I guess. (i.e. an add mutation should not allow a value to go backwards.) 


In my case for others that might find this. . 

Mutation A: 1
Mutation B: 2
Mutation C: 3

produces Layers (inMemoryCache uses layers of data)
LC: {defensiveShots: 3} <-current app facing value
LB: {defensiveShots: 2}
LA: {defensiveShots: 1}
Root 

If mutation C returns it resolves LC and leaves LB as the Top layer causing the page to re-render with 2. then when the server returns the final value it renders back to 3 as the Root. 

Basically if I can serialize them out and back from the server things are resolved. It's when they return out of order that things are mistaken. Also mutations that modify the same type can exhibit similar behavior if they are not serialized.
